src/ralph/backends/database/clickhouse.py

```diff
@@ -0,0 +1,358 @@
+            database=self.database,
+            settings=client_options,
+        )
+
+    def status(self) -> DatabaseStatus:
+        """Checks ClickHouse connection status."""
+        try:
+            self.client.query("SELECT 1")
+        except ClickHouseError:
+            return DatabaseStatus.AWAY
+
+        return DatabaseStatus.OK
+
+    @enforce_query_checks
+    def get(self, query: ClickHouseQuery = None, chunk_size: int = 500):
+        """Gets table rows and yields them."""
+        fields = ",".join(query.return_fields) if query.return_fields else "event"
+
+        sql = f"SELECT {fields} FROM {self.event_table_name}"
+
+        if query.where_clause:
+            sql += f"  WHERE {query.where_clause}"
+
+        result = self.client.query(sql).named_results()
+
+        for statement in result:
+            yield statement
+
+    @staticmethod
+    def _validate_and_parse(statement: dict) -> Tuple[uuid.UUID, datetime.datetime]:
+        """Validate and parse required statement fields."""
+        timestamp_parser = isoparser(sep="T")
+
+        try:
+            # Validate that the statement id is a UUID
+            event_id = uuid.UUID(statement["id"], version=4)
+
+            # Convert to a timestamp object that ClickHouse can accept
+            emission_time = timestamp_parser.isoparse(statement["timestamp"])
+
+            # If the emission time didn't include a timezone, set to UTC. This should
+            # be configurable?
+            if (
+                emission_time.tzinfo is None
+                or emission_time.tzinfo.utcoffset(emission_time) is None
+            ):
+                emission_time = emission_time.replace(tzinfo=pytz.utc)
+
+            return event_id, emission_time
+        except KeyError as err:
+            raise BadFormatException(
+                f"Statement is missing id or timestamp field: " f"{statement}"
+            ) from err
+        except ValueError as err:
+            raise BadFormatException(
+                f"Statement has an invalid 'id' or 'timestamp' field: " f"{statement}"
+            ) from err
+
+    @staticmethod
+    def to_documents(
+        stream: Union[TextIO, List], ignore_errors: bool = False
+    ) -> Generator[dict, None, None]:
+        """Converts `stream` lines (one statement per line) to insert tuples."""
+        for line in stream:
+            statement = json.loads(line) if isinstance(line, str) else line
+
+            try:
+                event_id, emission_time = ClickHouseDatabase._validate_and_parse(
+                    statement
+                )
+
+                document = (event_id, emission_time, statement, json.dumps(statement))
+
+                yield document
+            except BadFormatException as exc:
+                if ignore_errors:
+                    logger.warning(str(exc))
+                    continue
+                raise
+
+    def bulk_import(self, batch: List, ignore_errors: bool = False) -> int:
```

Comment:
Does `batch` require to be a list or any iterable is a valid type for insertion?